WorkspaceServer/Packaging/ProjectAsset.cs

```diff
@@ -0,0 +1,254 @@
+        {
+            if (result != null)
+            {
+                return result;
+            }
+
+            FileStream fileStream = null;
+            try
+            {
+                fileStream = File.Create(_lockFile.FullName, 1, FileOptions.DeleteOnClose);
+                using (var operation = Log.OnEnterAndConfirmOnExit())
+                {
+                    try
+                    {
+                        operation.Info("Attempting building package {name}", Name);
+                        await DotnetBuild();
+                        operation.Info("Workspace built");
+                        operation.Succeed();
+                    }
+                    catch (Exception exception)
+                    {
+                        operation.Error("Exception building workspace", exception);
+                        throw;
+                    }
+
+                    var binLog = this.FindLatestBinLog();
+
+                    if (binLog == null)
+                    {
+                        throw new InvalidOperationException("Failed to build");
+                    }
+
+                    var results = await TryLoadAnalyzerResultsAsync(binLog);
+
+                    if (results?.Count == 0)
+                    {
+                        throw new InvalidOperationException("The build log seems to contain no solutions or projects");
+                    }
+
+                    result = results?.FirstOrDefault(p => p.ProjectFilePath == _projectFile.FullName);
+                    if (result != null)
+                    {
+                        if (result.Succeeded)
+                        {
+                            return result;
+                        }
+
+                        throw new InvalidOperationException("Failed to build");
+                    }
+
+                    throw new InvalidOperationException("Failed to build");
+                }
+            }
+            finally
+            {
+                fileStream?.Dispose();
+            }
+        }
+
+        private async Task<AnalyzerResults> TryLoadAnalyzerResultsAsync(FileInfo binLog)
+        {
+            AnalyzerResults results = null;
+            await binLog.DoWhenFileAvailable(() =>
+            {
+                var manager = new AnalyzerManager();
+                results = manager.Analyze(binLog.FullName);
+            });
+            return results;
+        }
+
+        public Task<Workspace> CreateRoslynWorkspaceAsync(Budget budget)
+        {
+            return _workspaceStep.GetLatestAsync();
+        }
+
+        public Task<Workspace> CreateRoslynWorkspaceForRunAsync(Budget budget)
+        {
+            return CreateRoslynWorkspaceAsync(budget);
+        }
+
+        public Task<Workspace> CreateRoslynWorkspaceForLanguageServicesAsync(Budget budget)
+        {
+            return CreateRoslynWorkspaceAsync(budget);
+        }
```

Comment:
Is CreateRoslynWorkspaceForRun and for language service intended to do the same thing ?

Reply:
Yes. The distinction between the two may not be meaningful.